src/Core/Addon/Module/ModuleManager.php

```diff
@@ -0,0 +1,273 @@
+     * or a location (url or path to the zip file)
+     * @return bool true for success
+     */
+    public function install($name)
+    {
+        if (!$this->employee->can('add', 'AdminModules')) {
+            throw new Exception('You are not allowed to install a module');
+        }
+
+        if ($this->moduleProvider->isInstalled($name)) {
+            throw new Exception(sprintf('The module %s is already installed', $name));
+        }
+
+        if (! $this->moduleProvider->isOnDisk($name)) {
+            $this->moduleUpdater->setModuleOnDiskFromAddons($name);
+        }
+
+        $module = $this->moduleRepository->getModule($name);
+        return $module->onInstall();
+    }
+
+    /**
+     * Remove all theme files, resources, documentation and specific modules
+     *
+     * @param string $name The source can be a module name (installed from either local disk or addons.prestashop.com).
+     * or a location (url or path to the zip file)
+     * @return bool true for success
+     */
+    public function uninstall($name, $file_deletion = false)
+    {
+        // Check permissions:
+        // * Employee can delete
+        // * Employee can delete this specific module
+        if (!$this->employee->can('delete', 'AdminModules')
+            || !$this->moduleProvider->can('uninstall', $name)) {
+            throw new Exception('You are not allowed to uninstall this module');
+        }
+
+        // Is module installed ?
+        if (! $this->moduleProvider->isInstalled($name)) {
+            return false;
+        }
+
+        // Get module instance and uninstall it
+        $module = $this->moduleRepository->getModule($name);
+        try {
+            $result = $module->onUninstall();
+        } catch (Exception $e) {
+            $this->msgs_error[$name][] = 'Error on module uninstallation. '. $e->getMessage();
+            $result = false;
+        }
+
+        if ($result && (bool)$file_deletion) {
+            $result &= $this->removeModuleFromDisk($name);
+        }
+
+        return $result;
+    }
+
+    /**
+    * Download new files from source, backup old files, replace files with new ones
+    * and execute all necessary migration scripts form current version to the new one.
+    *
+    * @param Addon $name the theme you want to upgrade
+    * @param string $version the version you want to up upgrade to
+    * @param string $source if the upgrade is not coming from addons, you need to specify the path to the zipball
+    * @return bool true for success
+    */
+    public function upgrade($name, $version = 'latest', $source = null)
+    {
+        if (!$this->employee->can('edit', 'AdminModules')
+            || !$this->moduleProvider->can('configure', $name)) {
+            throw new Exception('You are not allowed to upgrade this module');
+        }
+
+        if (! $this->moduleProvider->isInstalled($name)) {
+            throw new Exception('This module must be installed');
+        }
+
+        // Get new module
```

Comment:
I'd like to see this implementation complete before merge. We have a lot (too much ?) code merged and still unfinished, this lead to a lot of debugging.

If you can't finish it, throw an `NotImplementedException` here please.